Interface Tree traversal

rmhasan opened the pull request:
Added following methods in ``InterfaceViewSet`` which can be invoked by get requests to return view sets

1. parent
2. ancestors
3. children
4. descendants
5. siblings
6. root

Also implemented following methods in ``Interface`` model class in ``models.py`` which get invoked in the view methods

1. get_ancestors
2. get_children
3. get_descendants
4. get_root
5. get_siblings

Also added tests in ``tests/api_tests/test_interfaces.py`` and ``tests/model_tests/test_interfaces.py``

jathanism commented:
Oh awesome. Sorry I've been behind the curve. I'll check this out. Thank you!!

jathanism commented:
I'm going to review this damn thing eventually. I need a clone.

jathanism commented on nsot/api/views.py:
```diff
@@ -670,25 +670,55 @@ def addresses(self, request, pk=None, site_pk=None, *args, **kwargs):
+        """Return parent of interface"""
+        interface = self.get_resource_object(pk, site_pk)
+        parent = [] if interface.parent is None else [interface.parent]
+        return self.list(request, queryset=parent, *args, **kwargs)
```
Parent should return only a single object vs. a list. See `NetworkViewSet.parent` for a good reference on how it's done for network objects.

jathanism commented on nsot/models.py:
```diff
@@ -1396,8 +1396,43 @@ def set_addresses(self, addresses, overwrite=False, partial=False):
+
+    def get_siblings(self):
+        """Return the interfaces with same parent as an interface"""
+        return list(Interface.objects.filter(parent=self.parent).exclude(id=self.id))
```
Siblings should be interfaces with both the same parent AND the same device, otherwise all interfaces with no parent are considered to be siblings, which I don't think is what we really want.

rmhasan replied:
@jathanism Currently you can create a interface on a device and set its parent to an interface on another device. Should I create a validation check that prevents that?

jathanism replied:
Yeah I guess that is a pretty sane constraint!

rmhasan replied:
Okay I will work on this.

nickpegg commented on nsot/models.py:
```diff
@@ -1396,8 +1396,43 @@ def set_addresses(self, addresses, overwrite=False, partial=False):
+        root = self.parent
+        while root is not None and root.parent is not None:
+            root = root.parent
+        return [] if root is None else [root]
```
Similar to @jathanism's [comment on `parent`](https://github.com/dropbox/nsot/pull/271/files#r116053169), this should have the same behavior and return a single object rather than a list.

rmhasan commented:
Hi @jathanism and @nickpegg, I saw the feed back on #266 and #271. I will work on them soon.

rmhasan commented:
Hi @jathanism and @nickpegg I worked on the changes you requested. Sorry it took so long.

jathanism commented:
@rmhasan There's no need to apologize, you're a hero!

rmhasan commented:
Hi @jathanism and @nickpegg. I made the following changes.

- use `self.retreive` to return parent and root interfaces
- change filter query for `siblings` view to look for interfaces with same parent and device
- return Interface whose parent is `None` as root if `root` is called on that interface
- also added validation check to prevent users from setting an interfaces parent to a interface
  with a different device
- added test to check when `siblings` is called, that interfaces with same parent and devices are returned
- added test to check when `root` is called on an interface with no parent that it is returned
- added test to check that a user cannot set an interface's parent to an interface with different device

jathanism commented on nsot/api/views.py:
```diff
@@ -662,33 +662,68 @@ def get_serializer_class(self):
 
     @detail_route(methods=['get'])
+    def parent(self, request, pk=None, site_pk=None, *args, **kwargs):
+        """Return parent of interface"""
```
Can you please update the docstring to be similar in style to the one at `nsot.api.NetworkViewSet.parent()`?

And for the remaining methods you've added can you please make sure they all have docstrings?

This is important because the docstrings are used in the automated API documentation on nsot.readthedocs.io and in the browsable API.

jathanism commented on nsot/models.py:
```diff
@@ -1396,8 +1396,43 @@ def set_addresses(self, addresses, overwrite=False, partial=False):
+    def get_ancestors(self):
+        """Recursively get all parents of an interface"""
+        p = self.parent
+        ancestors = []
```
Just a comment: Returning a list instead of a queryset may be problematic if we attempt to implement any filtering (or filter chaining) for ancestors. This isn't currently a use-case but it will be something to keep in mind if the need arises in the future.

rmhasan replied:
Hi @jathanism, how about I do the following. I get all the id's of the ancestors in a list like the following 

`ancestor_ids = [a.id for a in ancestors]`

Then I return the `Queryset` from the following query

`Interface.objects.filter(id__in=ancestor_ids)`

jathanism commented on nsot/models.py:
```diff
@@ -1396,8 +1396,43 @@ def set_addresses(self, addresses, overwrite=False, partial=False):
+    def get_descendants(self):
+        """Recursively return all the children of an interface"""
+        s = list(self.get_children())
+        children = []
```
Just a comment: Similarly to ancestors, returning a list may be problematic if we ever need to filter (or chain filters) on descendant interfaces.

rmhasan replied:
I would use the same fix mentioned in the `Interface.get_ancestors` comment.

jathanism commented on nsot/models.py:
```diff
@@ -1396,8 +1396,43 @@ def set_addresses(self, addresses, overwrite=False, partial=False):
+
+    def get_siblings(self):
+        """Return the interfaces with same parent as an interface"""
+        return list(Interface.objects.filter(parent=self.parent, device=self.device).exclude(id=self.id))
```
Just a comment: Last but not least, casting the queryset to a list may become problematic in the future in the event we need to implement filtering or desire to chain filters on these results.

rmhasan commented:
Hi @jathanism  I made the following changes
- I modified the docstrings of some functions and added docstrings to others
- I changed the message in the `ValidationError` thrown from `Interface.clean_parent` to be a  double quoted string
- Instead of returning lists, I'm returning `Querysets` from `Interface.get_descendants, Interface.get_ancestors, and Interface.get_siblings`

nickpegg commented:
Thanks for this contribution! :) I'm merging this and cutting a new version. I'll drop a note on this PR once that version's available on PyPI.

nickpegg commented:
v.1.1.4 released to PyPI: https://pypi.python.org/pypi/nsot/1.1.4
